Flaky test TestTBufferedServer_Metrics

it strikes again https://travis-ci.com/github/yurishkuro/jaeger/jobs/440719322#L679
```
=== RUN   TestTBufferedServer_Metrics
    metricstest.go:49: 
        	Error Trace:	metricstest.go:49
        	            				metricstest.go:37
        	            				tbuffered_server_test.go:156
        	Error:      	Not equal: 
        	            	expected: int(1)
        	            	actual  : int64(0)
        	Test:       	TestTBufferedServer_Metrics
        	Messages:   	expected metric name: thrift.udp.server.read.errors, tags: map[]
--- FAIL: TestTBufferedServer_Metrics (0.01s)
FAIL
coverage: 97.3% of statements
FAIL	github.com/jaegertracing/jaeger/cmd/agent/app/servers	0.051s
```

--- a/cmd/agent/app/servers/tbuffered_server_test.go
+++ b/cmd/agent/app/servers/tbuffered_server_test.go
@@ -98,11 +98,13 @@ type fakeTransport struct {
 
 func (t *fakeTransport) Read(p []byte) (n int, err error) {
 	packet := t.packet.Inc()
-	if packet > 2 {
-		if packet > 3 {
-			// return error once when packet==3, otherwise block
-			t.wg.Wait()
-		}
+	if packet == 2 {
+		// return some error packet, followed by valid one
+		return 0, io.ErrNoProgress
+	}
+	if packet > 3 {
+		// don't return error at this moment to avoid flaky test
+		t.wg.Wait()
 		return 0, io.EOF
 	}
 	for i := range p {
@@ -128,9 +130,9 @@ func TestTBufferedServer_Metrics(t *testing.T) {
 	go server.Serve()
 	defer server.Stop()
 
-	// The fakeTransport will allow the server to read exactly two packets and one error.
+	// The fakeTransport will allow the server to read exactly two packets and one error in between.
 	// Since we use the server with queue size == 1, the first packet will be
-	// sent to channel, and the second one dropped.
+	// sent to channel, the error will be processed and the second valid packet dropped.
 
 	packetDropped := false
 	for i := 0; i < 5000; i++ {

--- a/cmd/agent/app/servers/tbuffered_server_test.go
+++ b/cmd/agent/app/servers/tbuffered_server_test.go
@@ -96,6 +96,10 @@ type fakeTransport struct {
 	wg     sync.WaitGroup
 }
 
+// Read simulates three packets received, then blocks until semaphore is released at the end of the test.
+// First packet is returned as normal.
+// Second packet is simulated as error.
+// Third packet is returned as normal, but will be dropped as overflow by the server whose queue size = 1.
 func (t *fakeTransport) Read(p []byte) (n int, err error) {
 	packet := t.packet.Inc()
 	if packet == 2 {
@@ -103,7 +107,7 @@ func (t *fakeTransport) Read(p []byte) (n int, err error) {
 		return 0, io.ErrNoProgress
 	}
 	if packet > 3 {
-		// don't return error at this moment to avoid flaky test
+		// block after 3 packets until the server is shutdown and semaphore released
 		t.wg.Wait()
 		return 0, io.EOF
 	}
@@ -132,7 +136,7 @@ func TestTBufferedServer_Metrics(t *testing.T) {
 
 	// The fakeTransport will allow the server to read exactly two packets and one error in between.
 	// Since we use the server with queue size == 1, the first packet will be
-	// sent to channel, the error will be processed and the second valid packet dropped.
+	// sent to channel, the error will increment the metric, and the second valid packet dropped.
 
 	packetDropped := false
 	for i := 0; i < 5000; i++ {

